### PR TITLE
Support BitVec, BitBox and BitSlice

### PR DESCRIPTION
**Pull request by DarkEld3r**
Naive implementation to start with. I suppose we also need `BitBox` and `BitSlice` support. Not sure if flexibility to use different [cursor](https://docs.rs/bitvec/0.11.1/bitvec/cursor/index.html) and inner types is needed.

Needed for https://github.com/paritytech/polkadot/issues/254.

### CODE REVIEW

File: src/codec.rs [hunk 1]
```diff
@@ -1227,6 +1230,29 @@ impl Decode for () {
 	}
 }
 
+#[cfg(feature = "bit-vec")]
+impl Encode for BitVec {
+	fn encode_to<W: Output>(&self, dest: &mut W) {
+		let len = self.len();
+		assert!(len <= u32::max_value() as usize, "Attempted to serialize a collection with too many elements.");
+		Compact(len as u32).encode_to(dest);
```
Comment (bkchr) [inline src/codec.rs]: Why do we encode the length twice?

File: src/codec.rs [hunk 2]
```diff
@@ -1227,6 +1230,29 @@ impl Decode for () {
...
+}
+
+#[cfg(feature = "bit-vec")]
+impl Decode for BitVec {
+	fn decode<I: Input>(input: &mut I) -> Result<Self, Error> {
+		<Compact<u32>>::decode(input).and_then(move |Compact(len)| {
+			let len = len as usize;
+			let mut v = Self::from_vec(Vec::<u8>::decode(input)?);
+			assert!(len <= v.len());
+			unsafe { v.set_len(len); }
```
Comment (bkchr) [inline src/codec.rs]: I think the length should already be correctly set?

File: src/codec.rs [hunk 3]
```diff
@@ -1901,4 +1925,27 @@ mod tests {
 			check_bound_high!(i, [(u128, U128_OUT_OF_RANGE)]);
 		}
 	}
+
+	#[cfg(feature = "bit-vec")]
```
Comment (bkchr) [inline src/codec.rs]: Please enable the feature in CI while testing.

File: src/codec.rs (hunk 1 above)
Comment (DarkEld3r) [reply src/codec.rs]: I agree that it is somewhat ugly, but these are different lengths. The first one is a number of bits in `BitVec` and the second one is the length of a slice in elements (bytes):
```rust
let vec = bitvec![0, 0];
assert_eq!(2, vec.len());
assert_eq!(1, vec.as_slice().len());
```

Comment (bkchr) [reply src/codec.rs]: Ahh okay, but as the length of slices is computable by the number of bits, you should just do 
`dest.write(self.as_slice())` to write without the second length.

**Comment by DarkEld3r** [comment]
Updated and ready for review.

File: src/bit_vec.rs [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+use std::mem;
+
+use bitvec::{vec::BitVec, bits::Bits, cursor::Cursor, slice::BitSlice, boxed::BitBox};
+use byte_slice_cast::{AsByteSlice, ToByteSlice, FromByteSlice, Error as FromByteSliceError};
+
+use crate::codec::{Encode, Decode, Input, Output, Compact, Error};
+
+impl From<FromByteSliceError> for Error {
+	fn from(_: FromByteSliceError) -> Error {
+		"failed to cast from byte slice".into()
```
[the remark this answers is not in the document]
Comment (DarkEld3r) [reply src/bit_vec.rs]: Thanks, fixed.

File: src/bit_vec.rs [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+
+use bitvec::{vec::BitVec, bits::Bits, cursor::Cursor, slice::BitSlice, boxed::BitBox};
+use byte_slice_cast::{AsByteSlice, ToByteSlice, FromByteSlice, Error as FromByteSliceError};
+
+use crate::codec::{Encode, Decode, Input, Output, Compact, Error};
+
+impl From<FromByteSliceError> for Error {
+	fn from(e: FromByteSliceError) -> Error {
+		match e {
+			FromByteSliceError::AlignmentMismatch{..} => "failed to cast from byte slice: alignment mismatch".into(),
```
Comment (thiolliere) [inline src/bit_vec.rs]:
```suggestion
			FromByteSliceError::AlignmentMismatch {..} => "failed to cast from byte slice: alignment mismatch".into(),
```

File: src/bit_vec.rs [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+			FromByteSliceError::LengthMismatch{..} => "failed to cast from byte slice: length mismatch".into(),
```
Comment (thiolliere) [inline src/bit_vec.rs]:
```suggestion
			FromByteSliceError::LengthMismatch {..} => "failed to cast from byte slice: length mismatch".into(),
```

File: src/bit_vec.rs (hunk 5 above)
Comment (DarkEld3r) [reply src/bit_vec.rs]: If only we used `rustfmt`... :upside_down_face:

Comment (thiolliere) [reply src/bit_vec.rs]: I think some IDE should be able to run rustfmt on a specific line range. but I coudln't manage to use it in vim :-/

**Pull request merged**